lxl-web/src/lib/utils/cslFromMainEntity.ts

```diff
@@ -50,29 +53,109 @@ export function cslFromMainEntity(mainEntity: FramedData, vocabUtil: VocabUtil):
-			constributors = jmespath.search(
-				mainEntity,
-				`instanceOf.contribution[?"@type"=='PrimaryContribution'].${agentPicker}`
-			);
-			if (!constributors?.length) {
-				constributors = jmespath.search(
-					mainEntity,
-					`instanceOf.contribution[?"@type"=='Contribution'].${agentPicker}`
-				);
-				result.author = constributors;
+	// for each contributor, assign a role
+	const contributors = mapContribution(mainEntity.instanceOf?.contribution as Contributor[]);
+	result = { ...result, ...contributors };
+	result.type = getCslType(mainEntity, vocabUtil);
+
+	return [result as CSLJSON];
+}
+
+function mapContribution(contribution: Contributor[]): Partial<CSLJSON> {
+	const result: Partial<CSLJSON> = {};
+
+	if (contribution && Array.isArray(contribution)) {
+		for (const contributor of contribution) {
+			const roleEntry = getRole(contributor);
+			if (roleEntry) {
+				for (const [role, name] of Object.entries(roleEntry) as [keyof CSLRoles, CSLName[]][]) {
+					addContributor(result, role, name);
+				}
+			} else {
+				const fallbackRole = getFallbackRole(contributor);
+				if (fallbackRole) {
+					addContributor(result, 'author', fallbackRole);
+				}
 			}
-		} catch {
-			// do nothing
 		}
 	}
+	return result;
+}
 
-	result.type = getCslType(mainEntity, vocabUtil);
+function addContributor(result: Partial<CSLJSON>, role: keyof CSLRoles, names: CSLName[]) {
+	if (Array.isArray(result[role])) {
+		result[role].push(...names);
+	} else {
+		result[role] = names;
+	}
+}
 
-	return [result as CSLJSON];
+function getRole(contributor: Contributor): Partial<CSLRoles> | false {
+	if (contributor.role) {
+		let role: string | undefined = contributor.role?.[0]?.['@id'];
+		if (role) {
+			role = role.replace('https://id.kb.se/relator/', '');
+			const name = getName(contributor.agent);
+			switch (role) {
+				case 'author':
+					return { author: name };
+				case 'editor':
+					return { editor: name };
+				case 'composer':
+					return { composer: name };
+				case 'filmDirector':
+				case 'director':
+					return { director: name };
+				case 'illustrator':
+					return { illustrator: name };
+				case 'interviewer':
+					return { interviewer: name };
+				case 'translator':
+					return { translator: name };
+				case 'producer':
+				case 'filmProducer':
+					return { producer: name };
+				case 'narrator':
+					return { narrator: name };
+				case 'performer':
+				case 'musician':
+				case 'singer':
+					return { performer: name };
+				case 'compiler':
+					return { compiler: name };
+				case 'curator':
+					return { curator: name };
+				default:
+					return false;
+			}
+		}
+	}
+	return false;
+}
+
+function getFallbackRole(contributor: Contributor) {
+	if (contributor['@type'] === 'PrimaryContribution' || contributor['@type'] === 'Contribution') {
+		const name = getName(contributor.agent);
+		if (name) {
+			return name;
+		}
+	}
+	return false;
+}
+
+function getName(agent: Agent | Agent[]): CSLName[] {
```

Comment:
list of properties to consider for person names...
https://github.com/libris/definitions/blob/0b8f45f88094badc5bbd7d87cd285663168c9c80/source/vocab/display.jsonld#L410

see e.g.
http://localhost:5173/3ld2dlqf2v2mbc6#work-record

Another Henrik II...
http://localhost:5173/h1t1r26t44g4wt2?_q=contributor%3A%22libris%3Amkz12s953slfm6t%23it%22&cite=h1t1r26t44g4wt2

https://sv.wikipedia.org/wiki/Henrik_II_av_Frankrike


I think it's ok to just check these when name is in `name` property